File: scripts/test-llm-assistant.js
/**
 * Test LLM Assistant API Endpoints
 * Tests page extraction, ask about page, and summarization
 */

const API_URL = process.env.VITE_API_URL || 'http://localhost:8000';

async function testPageExtraction() {
  console.log('\n📄 Testing Page Extraction...');
  try {
    const response = await fetch(`${API_URL}/extract/extract`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        url: 'https://en.wikipedia.org/wiki/TypeScript',
      }),
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const data = await response.json();
    console.log('✅ Page extraction successful');
    console.log(`   Title: ${data.title}`);
    console.log(`   Content length: ${data.content.length} chars`);
    console.log(`   Language: ${data.lang}`);
    return true;
  } catch (error) {
    console.error('❌ Page extraction failed:', error.message);
    return false;
  }
}

async function testAskAboutPage() {
  console.log('\n❓ Testing Ask About Page...');
  try {
    const response = await fetch(`${API_URL}/llm/ask-about-page`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        prompt: 'What is TypeScript?',
        url: 'https://en.wikipedia.org/wiki/TypeScript',
      }),
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    // Handle SSE stream
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let answer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          try {
            const data = JSON.parse(line.slice(6));
            if (data.type === 'token' && data.text) {
              answer += data.text;
            }
          } catch (e) {
            // Skip invalid JSON
          }
        }
      }
    }

    console.log('✅ Ask about page successful');
    console.log(`   Answer length: ${answer.length} chars`);
    console.log(`   Preview: ${answer.substring(0, 100)}...`);
    return true;
  } catch (error) {
    console.error('❌ Ask about page failed:', error.message);
    return false;
  }
}

async function testPageSummarization() {
  console.log('\n📝 Testing Page Summarization...');
  try {
    const response = await fetch(`${API_URL}/llm/summarize-page`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        url: 'https://en.wikipedia.org/wiki/TypeScript',
        style: 'concise',
        max_length: 100,
      }),
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    // Handle SSE stream
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    let summary = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          try {
            const data = JSON.parse(line.slice(6));
            if (data.type === 'token' && data.text) {
              summary += data.text;
            }
          } catch (e) {
            // Skip invalid JSON
          }
        }
      }
    }

    console.log('✅ Page summarization successful');
    console.log(`   Summary length: ${summary.length} chars`);
    console.log(`   Summary: ${summary.substring(0, 200)}...`);
    return true;
  } catch (error) {
    console.error('❌ Page summarization failed:', error.message);
    return false;
  }
}

async function testLLMAssistant() {
  console.log('\n🤖 Testing LLM Assistant...');
  try {
    const response = await fetch(`${API_URL}/llm/assistant`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        prompt: 'Explain quantum computing in simple terms',
        context: 'Quantum computing uses quantum mechanics principles',
        stream: false,
      }),
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const data = await response.json();
    console.log('✅ LLM Assistant successful');
    console.log(`   Model: ${data.model}`);
    console.log(`   Response length: ${data.response.length} chars`);
    console.log(`   Preview: ${data.response.substring(0, 100)}...`);
    return true;
  } catch (error) {
    console.error('❌ LLM Assistant failed:', error.message);
    return false;
  }
}

async function testAISearch() {
  console.log('\n🔍 Testing AI Search...');
  try {
    const response = await fetch(`${API_URL}/search/ai-search`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        query: 'React performance optimization',
        max_results: 10,
        include_summary: true,
        stream: false,
      }),
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const data = await response.json();
    console.log('✅ AI Search successful');
    console.log(`   Total results: ${data.total_results}`);
    console.log(`   Summary length: ${data.summary?.length || 0} chars`);
    if (data.summary) {
      console.log(`   Summary preview: ${data.summary.substring(0, 100)}...`);
    }
    return true;
  } catch (error) {
    console.error('❌ AI Search failed:', error.message);
    return false;
  }
}

async function testSearchWithSummary() {
  console.log('\n🔎 Testing Search with Summary...');
  try {
    const response = await fetch(`${API_URL}/search`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        query: 'TypeScript best practices',
        max_results: 15,
        include_summary: true,
        summary_length: 150,
      }),
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    const data = await response.json();
    console.log('✅ Search with summary successful');
    console.log(`   Total results: ${data.total_results}`);
    console.log(`   Sources used: ${data.sources_used.join(', ')}`);
    if (data.summary) {
      console.log(`   Summary: ${data.summary.substring(0, 150)}...`);
    }
    return true;
  } catch (error) {
    console.error('❌ Search with summary failed:', error.message);
    return false;
  }
}

async function runAllTests() {
  console.log('🧪 Starting LLM Assistant & AI Search Tests\n');
  console.log(`API URL: ${API_URL}\n`);

  const results = {
    pageExtraction: await testPageExtraction(),
    askAboutPage: await testAskAboutPage(),
    pageSummarization: await testPageSummarization(),
    llmAssistant: await testLLMAssistant(),
    aiSearch: await testAISearch(),
    searchWithSummary: await testSearchWithSummary(),
  };

  const passed = Object.values(results).filter(Boolean).length;
  const total = Object.keys(results).length;

  console.log('\n' + '='.repeat(50));
  console.log('📊 Test Results Summary');
  console.log('='.repeat(50));
  console.log(`✅ Passed: ${passed}/${total}`);
  console.log(`❌ Failed: ${total - passed}/${total}`);
  console.log('\nDetailed Results:');
  for (const [test, result] of Object.entries(results)) {
    console.log(`  ${result ? '✅' : '❌'} ${test}`);
  }

  if (passed === total) {
    console.log('\n🎉 All tests passed!');
    process.exit(0);
  } else {
    console.log('\n⚠️  Some tests failed');
    process.exit(1);
  }
}

// Run tests
runAllTests().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: apps/api/test_llm_assistant.py
"""
Test LLM Assistant API Endpoints
"""

import asyncio
import httpx
import json
import sys

API_URL = "http://localhost:8000"


async def test_page_extraction():
    """Test page content extraction"""
    print("\n📄 Testing Page Extraction...")
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(
                f"{API_URL}/extract/extract",
                json={"url": "https://en.wikipedia.org/wiki/TypeScript"},
            )
            response.raise_for_status()
            data = response.json()
            print("✅ Page extraction successful")
            print(f"   Title: {data.get('title', 'N/A')}")
            print(f"   Content length: {len(data.get('content', ''))} chars")
            print(f"   Language: {data.get('lang', 'N/A')}")
            return True
    except Exception as e:
        print(f"❌ Page extraction failed: {e}")
        return False


async def test_ask_about_page():
    """Test ask about page endpoint"""
    print("\n❓ Testing Ask About Page...")
    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                f"{API_URL}/llm/ask-about-page",
                json={
                    "prompt": "What is TypeScript?",
                    "url": "https://en.wikipedia.org/wiki/TypeScript",
                },
            )
            response.raise_for_status()
            
            # Handle SSE stream
            answer = ""
            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    try:
                        data = json.loads(line[6:])
                        if data.get("type") == "token" and data.get("text"):
                            answer += data["text"]
                        elif data.get("type") == "done":
                            break
                    except:
                        pass
            
            print("✅ Ask about page successful")
            print(f"   Answer length: {len(answer)} chars")
            if answer:
                print(f"   Preview: {answer[:100]}...")
            return True
    except Exception as e:
        print(f"❌ Ask about page failed: {e}")
        return False


async def test_page_summarization():
    """Test page summarization endpoint"""
    print("\n📝 Testing Page Summarization...")
    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                f"{API_URL}/llm/summarize-page",
                json={
                    "url": "https://en.wikipedia.org/wiki/TypeScript",
                    "style": "concise",
                    "max_length": 100,
                },
            )
            response.raise_for_status()
            
            # Handle SSE stream
            summary = ""
            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    try:
                        data = json.loads(line[6:])
                        if data.get("type") == "token" and data.get("text"):
                            summary += data["text"]
                        elif data.get("type") == "done":
                            break
                    except:
                        pass
            
            print("✅ Page summarization successful")
            print(f"   Summary length: {len(summary)} chars")
            if summary:
                print(f"   Summary: {summary[:200]}...")
            return True
    except Exception as e:
        print(f"❌ Page summarization failed: {e}")
        return False


async def test_llm_assistant():
    """Test general LLM assistant endpoint"""
    print("\n🤖 Testing LLM Assistant...")
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(
                f"{API_URL}/llm/assistant",
                json={
                    "prompt": "Explain quantum computing in simple terms",
                    "context": "Quantum computing uses quantum mechanics principles",
                    "stream": False,
                },
            )
            response.raise_for_status()
            data = response.json()
            print("✅ LLM Assistant successful")
            print(f"   Model: {data.get('model', 'N/A')}")
            print(f"   Response length: {len(data.get('response', ''))} chars")
            if data.get("response"):
                print(f"   Preview: {data['response'][:100]}...")
            return True
    except Exception as e:
        print(f"❌ LLM Assistant failed: {e}")
        return False


async def test_ai_search():
    """Test AI search endpoint"""
    print("\n🔍 Testing AI Search...")
    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                f"{API_URL}/search/ai-search",
                json={
                    "query": "React performance optimization",
                    "max_results": 10,
                    "include_summary": True,
                    "stream": False,
                },
            )
            response.raise_for_status()
            data = response.json()
            print("✅ AI Search successful")
            print(f"   Total results: {data.get('total_results', 0)}")
            if data.get("summary"):
                print(f"   Summary length: {len(data['summary'])} chars")
                print(f"   Summary preview: {data['summary'][:100]}...")
            return True
    except Exception as e:
        print(f"❌ AI Search failed: {e}")
        return False


async def test_search_with_summary():
    """Test search endpoint with summary"""
    print("\n🔎 Testing Search with Summary...")
    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                f"{API_URL}/search",
                json={
                    "query": "TypeScript best practices",
                    "max_results": 15,
                    "include_summary": True,
                    "summary_length": 150,
                },
            )
            response.raise_for_status()
            data = response.json()
            print("✅ Search with summary successful")
            print(f"   Total results: {data.get('total_results', 0)}")
            print(f"   Sources used: {', '.join(data.get('sources_used', []))}")
            if data.get("summary"):
                print(f"   Summary: {data['summary'][:150]}...")
            return True
    except Exception as e:
        print(f"❌ Search with summary failed: {e}")
        return False


async def run_all_tests():
    """Run all tests"""
    print("🧪 Starting LLM Assistant & AI Search Tests")
    print(f"API URL: {API_URL}\n")

    results = {
        "page_extraction": await test_page_extraction(),
        "ask_about_page": await test_ask_about_page(),
        "page_summarization": await test_page_summarization(),
        "llm_assistant": await test_llm_assistant(),
        "ai_search": await test_ai_search(),
        "search_with_summary": await test_search_with_summary(),
    }

    passed = sum(1 for v in results.values() if v)
    total = len(results)

    print("\n" + "=" * 50)
    print("📊 Test Results Summary")
    print("=" * 50)
    print(f"✅ Passed: {passed}/{total}")
    print(f"❌ Failed: {total - passed}/{total}")
    print("\nDetailed Results:")
    for test, result in results.items():
        print(f"  {'✅' if result else '❌'} {test}")

    if passed == total:
        print("\n🎉 All tests passed!")
        return 0
    else:
        print("\n⚠️  Some tests failed")
        return 1


if __name__ == "__main__":
    exit_code = asyncio.run(run_all_tests())
    sys.exit(exit_code)




